# Exercice 1 :


In [170]:
import pandas as pd

sports = ['Football','HandBall','SnowSport']

In [171]:
s = pd.Series(sports)
display(s)


0     Football
1     HandBall
2    SnowSport
dtype: object

In [172]:
print(s.dtypes)

object


In [173]:
numeric=[None, 15,100]
n = pd.Series(numeric)

sports = ['BasketBall','HandBall', None]
s = pd.Series(sports)

display(s)
display(n)

0    BasketBall
1      HandBall
2          None
dtype: object

0      NaN
1     15.0
2    100.0
dtype: float64

Dans la series contenant des nombre la None value s'affiche NaN tandis que dans celles avec les objects elle s'affiche None

In [174]:
import numpy as np
np.isnan(np.nan)
sports = ['BasketBall', 'HandBall', np.nan]
pd.Series(sports)


0    BasketBall
1      HandBall
2           NaN
dtype: object

In [175]:
sports = ['BasketBall', 'HandBall', 'Snowsport','baseBall', 'Swimming']

s = pd.Series(sports, index=['sport1','sport2','sport3','sport4','sport5'])

s.index



Index(['sport1', 'sport2', 'sport3', 'sport4', 'sport5'], dtype='object')

# Exercice 2 :

In [176]:
import numpy as np

np.isnan(np.nan)
sports={'bask': 'BasketBall', 'hand': 'HandBall', 'snow': 'Snowsport', 'base': 'baseBall','swim': 'Swimming'}

s=pd.Series(sports)

display(s)
s.index


bask    BasketBall
hand      HandBall
snow     Snowsport
base      baseBall
swim      Swimming
dtype: object

Index(['bask', 'hand', 'snow', 'base', 'swim'], dtype='object')

In [177]:
second_element = s.iloc[1]
print("Second element using iloc:", second_element)

swim_element = s.loc['swim']
print("Element with 'swim' label using loc:", swim_element)

Second element using iloc: HandBall
Element with 'swim' label using loc: Swimming


In [178]:
uppercase_series = s.str.upper()
print("Series with uppercase values:\n", uppercase_series)


Series with uppercase values:
 bask    BASKETBALL
hand      HANDBALL
snow     SNOWSPORT
base      BASEBALL
swim      SWIMMING
dtype: object


In [179]:
for element in sports:
  sports[element] = sports[element].upper()
print(sports)


#Autre facon plus technique de faire la loop :

#uppercase_sports = {key: value.upper() for key, value in sports.items()}
#print("Dictionary with uppercase values:\n", uppercase_sports)





{'bask': 'BASKETBALL', 'hand': 'HANDBALL', 'snow': 'SNOWSPORT', 'base': 'BASEBALL', 'swim': 'SWIMMING'}


In [180]:
import timeit

# Timeit setup for vectorized method
setup_vectorized = """
import pandas as pd
sports = {'bask': 'BasketBall', 'hand': 'HandBall', 'snow': 'Snowsport', 'base': 'baseBall', 'swim': 'Swimming'}
sIndex = pd.Series(sports)
"""

# Timeit setup for loop method
setup_loop = """
sports = {'bask': 'BasketBall', 'hand': 'HandBall', 'snow': 'Snowsport', 'base': 'baseBall', 'swim': 'Swimming'}
uppercase_sports = {key: value.upper() for key, value in sports.items()}
"""

# Timing the vectorized method
vectorized_time = timeit.timeit("sIndex.str.upper()", setup=setup_vectorized, number=10000)
print("Vectorized method time:", vectorized_time)

# Timing the loop method
loop_time = timeit.timeit("uppercase_sports = {key: value.upper() for key, value in sports.items()}", setup=setup_loop, number=10000)
print("Loop method time:", loop_time)


Vectorized method time: 0.9028379000010318
Loop method time: 0.0067918999993707985


La loop est plus rapide

In [181]:
#Solution 1 avec np.mean()
numeric=[None, 15,100]

num=pd.Series(numeric)
meanValue=np.mean(num)
print(meanValue)


#Solution 2 avec une loop
meanValue = 0
sum=0
i = 0
for item in num:
  if pd.notna(item):
    sum = sum + item
  else:
    i = i + 1 #compte les nan values

meanValue=sum/(len(num)-i)
print(meanValue)

57.5
57.5


In [182]:
import numpy as np
import pandas as pd
import timeit


numeric = [None, 15, 100]
num = pd.Series(numeric)

# Solution 1 avec np.mean()
def mean_vectorized():
    return np.mean(num)

# Solution 2 avec une loop
def mean_loop():
    sum_val = 0
    count = 0
    for item in num:
        if pd.notna(item):  
            sum_val += item
            count += 1
    return sum_val / count

# Time both methods using timeit
vectorized_time_mean = timeit.timeit("mean_vectorized()", globals=globals(), number=10000)
loop_time_mean = timeit.timeit("mean_loop()", globals=globals(), number=10000)

print("Vectorized mean method time:", vectorized_time_mean)
print("Loop mean method time:", loop_time_mean)


Vectorized mean method time: 0.6335275000019465
Loop mean method time: 0.02579229999537347


#  Exercice 3 :

In [183]:
pearson_Data =pd.read_csv('Pearsons.csv', delimiter= ',') #pas forcement besoin de mettre le delimiter

In [184]:
pearson_Data.shape

(10000, 15)

In [185]:
pearson_Data.head()

name  \
0                                       Deonte Stark   
1                                     Faustino Boyer   
2  Eddy Bogisich,33431 Dollie Squares Apt. 654,Po...   
3                                     Mervyn Kreiger   
4  Katlyn Doyle,4650 Beer Crossing Suite 848,Nort...   

                        address             city           state         zip  \
0            278 Mueller Plains       North Euna         Alabama  03404-4384   
1  70244 Skiles Falls Suite 030  North Altohaven      California  01522-1310   
2                           NaN              NaN             NaN         NaN   
3            376 Dorinda Stream     Shaniquafort  South Carolina  39347-4438   
4                           NaN              NaN             NaN         NaN   

                 phone                          email           work  \
0   (180)940-9676x4495           shanna73@hotmail.com     Hahn-Mayer   
1  (308)699-6239x81011            ferrell81@gmail.com  Buckridge Inc   
2                  NaN                            NaN            NaN   
3         869-985-6299  emmerich.griselda@hotmail.com    Witting PLC   
4                  NaN                            NaN            NaN   

                     work address       work city work state work zipcode  \
0  8177 Weber Throughway Apt. 341        Jaronton      Maine   51589-1424   
1              236 Kessler Center  New Gavynshire   Missouri   52234-6972   
2                             NaN             NaN        NaN          NaN   
3               521 Kemmer Manors        Nerytown   Kentucky        68774   
4                             NaN             NaN        NaN          NaN   

          work phone                 work email   account created on  
0        01240240340  heller.kirstin@glover.com  2001-09-06 06:15:24  
1  (486)896-6855x446   esta.dicki@bechtelar.com           1983-04-25  
2                NaN                        NaN                  NaN  
3      (212)169-8190        greyson39@purdy.com  1971-04-27 14:05:06  
4                NaN                        NaN                  NaN

In [186]:
pearson_Data.columns

Index(['name', 'address', 'city', 'state', 'zip', 'phone', 'email', 'work',
       'work address', 'work city', 'work state', 'work zipcode', 'work phone',
       'work email', 'account created on'],
      dtype='object')

# Exercice 4 :

In [187]:
pearson_Data.isnull()

name  address   city  state    zip  phone  email   work  work address  \
0     False    False  False  False  False  False  False  False         False   
1     False    False  False  False  False  False  False  False         False   
2     False     True   True   True   True   True   True   True          True   
3     False    False  False  False  False  False  False  False         False   
4     False     True   True   True   True   True   True   True          True   
...     ...      ...    ...    ...    ...    ...    ...    ...           ...   
9995  False    False  False  False  False  False  False  False         False   
9996  False    False  False  False  False  False  False  False         False   
9997  False    False  False  False  False  False  False  False         False   
9998  False    False  False  False  False  False  False  False          True   
9999  False     True   True   True   True   True   True   True          True   

      work city  work state  work zipcode  work phone  work email  \
0         False       False         False       False       False   
1         False       False         False       False       False   
2          True        True          True        True        True   
3         False       False         False       False       False   
4          True        True          True        True        True   
...         ...         ...           ...         ...         ...   
9995      False       False         False       False       False   
9996      False       False         False       False       False   
9997      False       False         False       False       False   
9998      False       False         False       False       False   
9999       True        True          True        True        True   

      account created on  
0                  False  
1                  False  
2                   True  
3                  False  
4                   True  
...                  ...  
9995               False  
9996               False  
9997               False  
9998               False  
9999                True  

[10000 rows x 15 columns]

In [188]:
pearson_Data.isnull().sum()


name                    81
address               3365
city                  3374
state                 3365
zip                   3362
phone                 3366
email                 3363
work                  3390
work address          3383
work city             3378
work state            3390
work zipcode          3378
work phone            3377
work email            3359
account created on    3368
dtype: int64

In [189]:
pearson_Data.isnull().sum().loc["name"]

81

In [190]:
pearson_Data[pearson_Data["name"].isnull()]

name                          address              city           state  \
148   NaN     558 Brycen Mission Suite 152        Cristmouth        Arkansas   
273   NaN       0481 Sanford Lake Apt. 439     Bashirianberg  North Carolina   
300   NaN    38689 Kimora Groves Suite 807    New Nadiahaven         Vermont   
330   NaN       077 Walsh Summit Suite 123        Rogahnfurt         Indiana   
467   NaN              87140 Loma Crescent   North Dixieport        Michigan   
...   ...                              ...               ...             ...   
9316  NaN               054 Aubrie Corners    East Genevieve   New Hampshire   
9487  NaN       9957 Rempel Wells Apt. 081         New Micky         Alabama   
9745  NaN  6277 Schneider Common Suite 939         Port Aura    North Dakota   
9816  NaN               986 Brianne Shoals  Port Wilbertstad     Mississippi   
9834  NaN             19233 Kreiger Meadow        Kleinville        Michigan   

             zip               phone                            email  \
148   52585-7480    +03(6)2449148729               corene08@gmail.com   
273   61948-9865  (967)713-7747x7329  hermiston.jenniffer@hotmail.com   
300   99133-2546  595.341.7775x50639               sfritsch@yahoo.com   
330        81369    +38(1)3931574807             tdenesik@hotmail.com   
467   32196-8397    +65(0)2946604451              lucille69@gmail.com   
...          ...                 ...                              ...   
9316       92371    296-242-9406x586         enrico.kuvalis@yahoo.com   
9487       24013        522-665-0735              rylan47@hotmail.com   
9745       10127        127.858.7550                 kacy29@gmail.com   
9816       99396    953.475.7297x373         ritchie.hadley@yahoo.com   
9834  42537-0382   (894)710-7706x071            hjalmer14@hotmail.com   

                   work                  work address         work city  \
148           Braun PLC    8795 Carissa Land Apt. 884  South Arachester   
273           Grant LLC                7375 Lynn Fork      Wuckerthaven   
300       Purdy-Farrell              29365 Nyah Flats     West Isabella   
330        Corwin-Fahey  717 Karol Stravenue Apt. 964      Trantowhaven   
467         Monahan Ltd         99220 Murphy Motorway     East Reyhaven   
...                 ...                           ...               ...   
9316      Schneider LLC          53056 Gaige Motorway    North Audriana   
9487  Bergnaum and Sons   60120 Klocko Alley Apt. 615       Adalineport   
9745         Becker PLC           186 Armstrong Lakes      East Rosalyn   
9816       Kautzer-Batz  22243 Feil Terrace Suite 469      New Catofurt   
9834        Beer-Ledner            48801 Bobbye Roads        West Tylor   

        work state work zipcode          work phone  \
148     New Mexico   68070-4086         09359769734   
273           Ohio   89342-4989        419-132-5389   
300        Florida        06342    +66(4)7117753075   
330           Utah   49873-6641   882.577.0021x2495   
467   Rhode Island        50641        773.245.8906   
...            ...          ...                 ...   
9316      Arkansas        73816        605.101.0879   
9487      Maryland   37182-8209   289.393.5802x9949   
9745          Ohio        07576  1-924-436-1559x850   
9816     Wisconsin        62178         09723145605   
9834      Arkansas        60338    849.567.9201x682   

                              work email   account created on  
148   hezzie.hettinger@mcglynnlarson.com           1999-12-12  
273                 colin.walsh@hane.com  1995-08-07 19:52:30  
300                     ocrooks@mohr.net           1975-04-07  
330          mschaden@gerholdschultz.net           03/10/2010  
467            parisian.willard@feil.org           04/29/1986  
...                                  ...                  ...  
9316         cephus.conroy@schroeder.com           01/21/2000  
9487                    mwolf@corwin.net           2000-09-07  
9745         kunde.carry@auer

In [191]:
pearson_Data.dropna(how='all', inplace=True)

In [192]:
pearson_Data.isnull().sum()

name                    81
address               3365
city                  3374
state                 3365
zip                   3362
phone                 3366
email                 3363
work                  3390
work address          3383
work city             3378
work state            3390
work zipcode          3378
work phone            3377
work email            3359
account created on    3368
dtype: int64

In [193]:
pearson_Data.dropna(inplace=True, axis=0)

In [194]:
pearson_Data.isnull().sum()

name                  0
address               0
city                  0
state                 0
zip                   0
phone                 0
email                 0
work                  0
work address          0
work city             0
work state            0
work zipcode          0
work phone            0
work email            0
account created on    0
dtype: int64

Si on  précise how=all il faut que toute la ligne ou colonne est des Nan values pour etre supp, ici aucunes colonnes ou ligne est entierement composé de Nan values.

• axis=0 or 1 allows to filter the missing values according to the row(axis=0) donc supprime les ligne or the column(axis=1) donc supprime les colonnes

• how=all allows to eliminate all rows or all columns having mainly missing values

• inplace = True allow to need the same dataframe without creating a new output

• thresh= integer value allows to keep only rows (or columns) having a missing value rate more than
thresh.

# Exercice 5 :

In [195]:
pearson_Data.dtypes

name                  object
address               object
city                  object
state                 object
zip                   object
phone                 object
email                 object
work                  object
work address          object
work city             object
work state            object
work zipcode          object
work phone            object
work email            object
account created on    object
dtype: object

In [196]:
pearson_Data[['First_name', 'Last_Name']] = pearson_Data['name'].str.split(' ', expand=True, n=1)
pearson_Data['Last_Name'].fillna('', inplace=True)  # Remplir les noms manquants avec une chaîne vide sinon erreur car pas meme nbr de lignes


In [197]:
pearson_Data

name                          address  \
0           Deonte Stark               278 Mueller Plains   
1         Faustino Boyer     70244 Skiles Falls Suite 030   
3         Mervyn Kreiger               376 Dorinda Stream   
5         Daquan Leffler      43142 Howell Ports Apt. 953   
6        Jace Konopelski       963 Senger Points Apt. 992   
...                  ...                              ...   
9991   Tawnya Williamson                 4160 Welch Roads   
9993       Cayla Gleason            505 Rogahn Trafficway   
9995        Ciera Wunsch    25198 Fernando Isle Suite 175   
9996         Beula Bauch               800 Keebler Stream   
9997   Garvin Runolfsson  4844 Dietrich Islands Suite 457   

                     city           state         zip                phone  \
0              North Euna         Alabama  03404-4384   (180)940-9676x4495   
1         North Altohaven      California  01522-1310  (308)699-6239x81011   
3            Shaniquafort  South Carolina  39347-4438         869-985-6299   
5              West Emory     Connecticut  08086-5571         874-576-0739   
6              Lake Fonda    North Dakota       15581   (083)182-4626x8012   
...                   ...             ...         ...                  ...   
9991           Amasahaven           Maine       76161  1-468-320-0246x8915   
9993        New Sabastian       Tennessee  56989-4233  (578)118-9244x85865   
9995  South Pricillashire        Colorado  53222-8926    005-769-2290x3662   
9996          East Denise           Maine       33693         933.047.0305   
9997             Auerport        Michigan       34497         312.046.9856   

                              email              work  \
0              shanna73@hotmail.com        Hahn-Mayer   
1               ferrell81@gmail.com     Buckridge Inc   
3     emmerich.griselda@hotmail.com       Witting PLC   
5           natosha.kutch@yahoo.com  Christiansen Ltd   
6      langosh.mathilde@hotmail.com       Glover-Howe   
...                             ...               ...   
9991             doreilly@yahoo.com    Schuster-Lemke   
9993                mtorp@gmail.com  Morissette Group   
9995       shannon.beatty@gmail.com          Metz-Kub   
9996     steuber.zakary@hotmail.com       Barrows Inc   
9997            nwehner@hotmail.com         Braun Inc   

                        work address         work city      work state  \
0     8177 Weber Throughway Apt. 341          Jaronton           Maine   
1                 236 Kessler Center    New Gavynshire        Missouri   
3                  521 Kemmer Manors          Nerytown        Kentucky   
5      3136 Jered Junction Suite 993       West Mallie   New Hampshire   
6               87986 Harvey Islands     South Zackary        Kentucky   
...                              ...               ...             ...   
9991              82853 Watsica View      South Jaydan         Georgia   
9993    484 Lubowitz Inlet Suite 138   East Sannieland  North Carolina   
9995           28254 Nikolaus Plains  West Emmitthaven       Minnesota   
9996   35083 Champlin Keys Suite 030       Port Yulisa    Rhode Island   
9997             7514 Samiyah Brooks          West Ham           Texas   

     work zipcode          work phone                         work email  \
0      51589-1424         01240240340          heller.kirstin@glover.com   
1      52234-6972   (486)896-6855x446           esta.dicki@bechtelar.com   
3           68774       (212)169-8190                greyson39@purdy.com   
5      40778-7891    215.605.7824x844         anderson.gifford@nader.com   
6      75232-9391       (270)125-2639            kassulke.fate@hills.com   
...           ...                 ...                                ...   
9991   54467-3704        264-477-5271             donell95@wilderman.com   
9993        66018  892-930-2586x31401               bina.pagac@johns.com   
9995   34939-5981   (154)316-2907x520          clemente.dach@leannon.com   
9996        3574

In [198]:
pearson_Date = pearson_Data[pearson_Data['Last_Name'].str.len() <= 16]
pearson_Date

name                        address                 city  \
0          Deonte Stark             278 Mueller Plains           North Euna   
1        Faustino Boyer   70244 Skiles Falls Suite 030      North Altohaven   
3        Mervyn Kreiger             376 Dorinda Stream         Shaniquafort   
5        Daquan Leffler    43142 Howell Ports Apt. 953           West Emory   
6       Jace Konopelski     963 Senger Points Apt. 992           Lake Fonda   
...                 ...                            ...                  ...   
9990        Luca Rogahn       3834 Amari Mews Apt. 548             Ivaville   
9991  Tawnya Williamson               4160 Welch Roads           Amasahaven   
9993      Cayla Gleason          505 Rogahn Trafficway        New Sabastian   
9995       Ciera Wunsch  25198 Fernando Isle Suite 175  South Pricillashire   
9996        Beula Bauch             800 Keebler Stream          East Denise   

               state         zip                phone  \
0            Alabama  03404-4384   (180)940-9676x4495   
1         California  01522-1310  (308)699-6239x81011   
3     South Carolina  39347-4438         869-985-6299   
5        Connecticut  08086-5571         874-576-0739   
6       North Dakota       15581   (083)182-4626x8012   
...              ...         ...                  ...   
9990       Minnesota       46241        (223)826-1211   
9991           Maine       76161  1-468-320-0246x8915   
9993       Tennessee  56989-4233  (578)118-9244x85865   
9995        Colorado  53222-8926    005-769-2290x3662   
9996           Maine       33693         933.047.0305   

                              email              work  \
0              shanna73@hotmail.com        Hahn-Mayer   
1               ferrell81@gmail.com     Buckridge Inc   
3     emmerich.griselda@hotmail.com       Witting PLC   
5           natosha.kutch@yahoo.com  Christiansen Ltd   
6      langosh.mathilde@hotmail.com       Glover-Howe   
...                             ...               ...   
9990           merri.okon@gmail.com       Weimann PLC   
9991             doreilly@yahoo.com    Schuster-Lemke   
9993                mtorp@gmail.com  Morissette Group   
9995       shannon.beatty@gmail.com          Metz-Kub   
9996     steuber.zakary@hotmail.com       Barrows Inc   

                        work address         work city      work state  \
0     8177 Weber Throughway Apt. 341          Jaronton           Maine   
1                 236 Kessler Center    New Gavynshire        Missouri   
3                  521 Kemmer Manors          Nerytown        Kentucky   
5      3136 Jered Junction Suite 993       West Mallie   New Hampshire   
6               87986 Harvey Islands     South Zackary        Kentucky   
...                              ...               ...             ...   
9990     3380 Adolfo Island Apt. 280       West Vernon        Illinois   
9991              82853 Watsica View      South Jaydan         Georgia   
9993    484 Lubowitz Inlet Suite 138   East Sannieland  North Carolina   
9995           28254 Nikolaus Plains  West Emmitthaven       Minnesota   
9996   35083 Champlin Keys Suite 030       Port Yulisa    Rhode Island   

     work zipcode          work phone                         work email  \
0      51589-1424         01240240340          heller.kirstin@glover.com   
1      52234-6972   (486)896-6855x446           esta.dicki@bechtelar.com   
3           68774       (212)169-8190                greyson39@purdy.com   
5      40778-7891    215.605.7824x844         anderson.gifford@nader.com   
6      75232-9391       (270)125-2639            kassulke.fate@hills.com   
...           ...                 ...                                ...   
9990        02955   135.182.7308x6079               tyson70@nikolaus.com   
9991   54467-3704        264-477-5271             donell95@wilderman.com   
9993        66018  892-930-2586x31401               bina.pagac@johns.com   
9995   34939-5981   (154)316-2907x520          clem

# Exercice 6 : 


### Olympics.csv

In [199]:
olympics_df = pd.read_csv('olympics.csv', delimiter= ',')
olympics_df.head()

0         1     2     3     4      5         6     7     8  \
0                NaN  № Summer  01 !  02 !  03 !  Total  № Winter  01 !  02 !   
1  Afghanistan (AFG)        13     0     0     2      2         0     0     0   
2      Algeria (ALG)        12     5     2     8     15         3     0     0   
3    Argentina (ARG)        23    18    24    28     70        18     0     0   
4      Armenia (ARM)         5     1     2     9     12         6     0     0   

      9     10       11    12    13    14              15  
0  03 !  Total  № Games  01 !  02 !  03 !  Combined total  
1     0      0       13     0     0     2               2  
2     0      0       15     5     2     8              15  
3     0      0       41    18    24    28              70  
4     0      0       11     1     2     9              12

On remarque que la première ligne pose problème et qu'il serait bien que la deuxième ligne devienne celle des colonnes

Nous allons donc redéclarer la dataframe en skippant la première ligne

In [200]:
olympics_df = pd.read_csv('olympics.csv', delimiter= ',',skiprows=1) #Je skip la première ligne pour avoir la deuxieme ligne comme header

In [201]:
olympics_df

Unnamed: 0  № Summer  01 !  02 !  03 !  \
0                               Afghanistan (AFG)        13     0     0     2   
1                                   Algeria (ALG)        12     5     2     8   
2                                 Argentina (ARG)        23    18    24    28   
3                                   Armenia (ARM)         5     1     2     9   
4                         Australasia (ANZ) [ANZ]         2     3     4     5   
..                                            ...       ...   ...   ...   ...   
142  Independent Olympic Participants (IOP) [IOP]         1     0     1     2   
143                            Zambia (ZAM) [ZAM]        12     0     1     1   
144                          Zimbabwe (ZIM) [ZIM]        12     3     4     1   
145                        Mixed team (ZZX) [ZZX]         3     8     5     4   
146                                        Totals        27  4809  4775  5130   

     Total  № Winter  01 !.1  02 !.1  03 !.1  Total.1  № Games  01 !.2  \
0        2         0       0       0       0        0       13       0   
1       15         3       0       0       0        0       15       5   
2       70        18       0       0       0        0       41      18   
3       12         6       0       0       0        0       11       1   
4       12         0       0       0       0        0        2       3   
..     ...       ...     ...     ...     ...      ...      ...     ...   
142      3         0       0       0       0        0        1       0   
143      2         0       0       0       0        0       12       0   
144      8         1       0       0       0        0       13       3   
145     17         0       0       0       0        0        3       8   
146  14714        22     959     958     948     2865       49    5768   

     02 !.2  03 !.2  Combined total  
0         0       2               2  
1         2       8              15  
2        24      28              70  
3         2       9              12  
4         4       5              12  
..      ...     ...             ...  
142       1       2               3  
143       1       1               2  
144       4       1               8  
145       5       4              17  
146    5733    6078           17579  

[147 rows x 16 columns]

In [202]:
olympics_df.dtypes #je verifie que les types correspondent bien

Unnamed: 0        object
№ Summer           int64
01 !               int64
02 !               int64
03 !               int64
Total              int64
№ Winter           int64
01 !.1             int64
02 !.1             int64
03 !.1             int64
Total.1            int64
№ Games            int64
01 !.2             int64
02 !.2             int64
03 !.2             int64
Combined total     int64
dtype: object

Les types sont cohérents

In [203]:
#Je remplace les colonnes par des noms plus explicites en accord avec les données

olympics_df.rename(columns={'Unnamed: 0':'Country','01 !': 'Summer Gold', '02 !': 'Summer Silver', '03 !': 'Summer Bronze','01 !.1': 'Winter Gold', '02 !.1': 'Winter Silver', '03 !.1': 'Winter Bronze','01 !.2': 'Total Gold', '02 !.2': 'Total Silver', '03 !.2': 'Total Bronze'}, inplace=True)
olympics_df


Country  № Summer  Summer Gold  \
0                               Afghanistan (AFG)        13            0   
1                                   Algeria (ALG)        12            5   
2                                 Argentina (ARG)        23           18   
3                                   Armenia (ARM)         5            1   
4                         Australasia (ANZ) [ANZ]         2            3   
..                                            ...       ...          ...   
142  Independent Olympic Participants (IOP) [IOP]         1            0   
143                            Zambia (ZAM) [ZAM]        12            0   
144                          Zimbabwe (ZIM) [ZIM]        12            3   
145                        Mixed team (ZZX) [ZZX]         3            8   
146                                        Totals        27         4809   

     Summer Silver  Summer Bronze  Total  № Winter  Winter Gold  \
0                0              2      2         0            0   
1                2              8     15         3            0   
2               24             28     70        18            0   
3                2              9     12         6            0   
4                4              5     12         0            0   
..             ...            ...    ...       ...          ...   
142              1              2      3         0            0   
143              1              1      2         0            0   
144              4              1      8         1            0   
145              5              4     17         0            0   
146           4775           5130  14714        22          959   

     Winter Silver  Winter Bronze  Total.1  № Games  Total Gold  Total Silver  \
0                0              0        0       13           0             0   
1                0              0        0       15           5             2   
2                0              0        0       41          18            24   
3                0              0        0       11           1             2   
4                0              0        0        2           3             4   
..             ...            ...      ...      ...         ...           ...   
142              0              0        0        1           0             1   
143              0              0        0       12           0             1   
144              0              0        0       13           3             4   
145              0              0        0        3           8             5   
146            958            948     2865       49        5768          5733   

     Total Bronze  Combined total  
0               2               2  
1               8              15  
2              28              70  
3               9              12  
4               5              12  
..            ...             ...  
142             2               3  
143             1               2  
144             1               8  
145             4              17  
146          6078           17579  

[147 rows x 16 columns]

Maintenant que nous avons bien établis les colonnes et les index nous allons nous pencher sur les data en elle même

In [204]:
olympics_df.isnull().sum() #je regarde si il y a des valeurs manquantes

Country           0
№ Summer          0
Summer Gold       0
Summer Silver     0
Summer Bronze     0
Total             0
№ Winter          0
Winter Gold       0
Winter Silver     0
Winter Bronze     0
Total.1           0
№ Games           0
Total Gold        0
Total Silver      0
Total Bronze      0
Combined total    0
dtype: int64

Pas de valeurs manquantes (quelle chance !)

In [205]:
olympics_df.info() #cela fait ce que j'ai fais précédemment (voir les types de données et les nan values) mais en une seul fois

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 147 entries, 0 to 146
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   Country         147 non-null    object
 1   № Summer        147 non-null    int64 
 2   Summer Gold     147 non-null    int64 
 3   Summer Silver   147 non-null    int64 
 4   Summer Bronze   147 non-null    int64 
 5   Total           147 non-null    int64 
 6   № Winter        147 non-null    int64 
 7   Winter Gold     147 non-null    int64 
 8   Winter Silver   147 non-null    int64 
 9   Winter Bronze   147 non-null    int64 
 10  Total.1         147 non-null    int64 
 11  № Games         147 non-null    int64 
 12  Total Gold      147 non-null    int64 
 13  Total Silver    147 non-null    int64 
 14  Total Bronze    147 non-null    int64 
 15  Combined total  147 non-null    int64 
dtypes: int64(15), object(1)
memory usage: 18.5+ KB


In [206]:
olympics_df.Country.unique()

array(['Afghanistan\xa0(AFG)', 'Algeria\xa0(ALG)', 'Argentina\xa0(ARG)',
       'Armenia\xa0(ARM)', 'Australasia\xa0(ANZ) [ANZ]',
       'Australia\xa0(AUS) [AUS] [Z]', 'Austria\xa0(AUT)',
       'Azerbaijan\xa0(AZE)', 'Bahamas\xa0(BAH)', 'Bahrain\xa0(BRN)',
       'Barbados\xa0(BAR) [BAR]', 'Belarus\xa0(BLR)', 'Belgium\xa0(BEL)',
       'Bermuda\xa0(BER)', 'Bohemia\xa0(BOH) [BOH] [Z]',
       'Botswana\xa0(BOT)', 'Brazil\xa0(BRA)',
       'British West Indies\xa0(BWI) [BWI]', 'Bulgaria\xa0(BUL) [H]',
       'Burundi\xa0(BDI)', 'Cameroon\xa0(CMR)', 'Canada\xa0(CAN)',
       'Chile\xa0(CHI) [I]', 'China\xa0(CHN) [CHN]', 'Colombia\xa0(COL)',
       'Costa Rica\xa0(CRC)', 'Ivory Coast\xa0(CIV) [CIV]',
       'Croatia\xa0(CRO)', 'Cuba\xa0(CUB) [Z]', 'Cyprus\xa0(CYP)',
       'Czech Republic\xa0(CZE) [CZE]', 'Czechoslovakia\xa0(TCH) [TCH]',
       'Denmark\xa0(DEN) [Z]', 'Djibouti\xa0(DJI) [B]',
       'Dominican Republic\xa0(DOM)', 'Ecuador\xa0(ECU)',
       'Egypt\xa0(EGY) [EGY] [Z]', 'Er

Je remarque que dans le nom des pays il y a des éléments "noisy data"

In [207]:
olympics_df.Country.nunique()

147

Je vais donc garder tout ce qui se trouve avant la première itération de "("

In [208]:
olympics_df['Country'] = olympics_df['Country'].apply(lambda x: x.split('(')[0].strip()) #.strip() enleve les espaces en trop
olympics_df.Country.unique()

array(['Afghanistan', 'Algeria', 'Argentina', 'Armenia', 'Australasia',
       'Australia', 'Austria', 'Azerbaijan', 'Bahamas', 'Bahrain',
       'Barbados', 'Belarus', 'Belgium', 'Bermuda', 'Bohemia', 'Botswana',
       'Brazil', 'British West Indies', 'Bulgaria', 'Burundi', 'Cameroon',
       'Canada', 'Chile', 'China', 'Colombia', 'Costa Rica',
       'Ivory Coast', 'Croatia', 'Cuba', 'Cyprus', 'Czech Republic',
       'Czechoslovakia', 'Denmark', 'Djibouti', 'Dominican Republic',
       'Ecuador', 'Egypt', 'Eritrea', 'Estonia', 'Ethiopia', 'Finland',
       'France', 'Gabon', 'Georgia', 'Germany', 'United Team of Germany',
       'East Germany', 'West Germany', 'Ghana', 'Great Britain', 'Greece',
       'Grenada', 'Guatemala', 'Guyana', 'Haiti', 'Hong Kong', 'Hungary',
       'Iceland', 'India', 'Indonesia', 'Iran', 'Iraq', 'Ireland',
       'Israel', 'Italy', 'Jamaica', 'Japan', 'Kazakhstan', 'Kenya',
       'North Korea', 'South Korea', 'Kuwait', 'Kyrgyzstan', 'Latvia',
       'L

In [209]:
olympics_df.Country.nunique() #je verifie si il tourjours le meme nombre de pays

147

In [210]:
olympics_df

Country  № Summer  Summer Gold  Summer Silver  \
0                         Afghanistan        13            0              0   
1                             Algeria        12            5              2   
2                           Argentina        23           18             24   
3                             Armenia         5            1              2   
4                         Australasia         2            3              4   
..                                ...       ...          ...            ...   
142  Independent Olympic Participants         1            0              1   
143                            Zambia        12            0              1   
144                          Zimbabwe        12            3              4   
145                        Mixed team         3            8              5   
146                            Totals        27         4809           4775   

     Summer Bronze  Total  № Winter  Winter Gold  Winter Silver  \
0                2      2         0            0              0   
1                8     15         3            0              0   
2               28     70        18            0              0   
3                9     12         6            0              0   
4                5     12         0            0              0   
..             ...    ...       ...          ...            ...   
142              2      3         0            0              0   
143              1      2         0            0              0   
144              1      8         1            0              0   
145              4     17         0            0              0   
146           5130  14714        22          959            958   

     Winter Bronze  Total.1  № Games  Total Gold  Total Silver  Total Bronze  \
0                0        0       13           0             0             2   
1                0        0       15           5             2             8   
2                0        0       41          18            24            28   
3                0        0       11           1             2             9   
4                0        0        2           3             4             5   
..             ...      ...      ...         ...           ...           ...   
142              0        0        1           0             1             2   
143              0        0       12           0             1             1   
144              0        0       13           3             4             1   
145              0        0        3           8             5             4   
146            948     2865       49        5768          5733          6078   

     Combined total  
0                 2  
1                15  
2                70  
3                12  
4                12  
..              ...  
142               3  
143               2  
144               8  
145              17  
146           17579  

[147 rows x 16 columns]

### fliker.csv

In [211]:
fliker = pd.read_csv('fliker.csv')
fliker.head()

Identifier             Edition Statement      Place of Publication  \
0         206                           NaN                    London   
1         216                           NaN  London; Virtue & Yorston   
2         218                           NaN                    London   
3         472                           NaN                    London   
4         480  A new edition, revised, etc.                    London   

  Date of Publication              Publisher  \
0         1879 [1878]       S. Tinsley & Co.   
1                1868           Virtue & Co.   
2                1869  Bradbury, Evans & Co.   
3                1851          James Darling   
4                1857   Wertheim & Macintosh   

                                               Title     Author  \
0                  Walter Forbes. [A novel.] By A. A      A. A.   
1  All for Greed. [A novel. The dedication signed...  A., A. A.   
2  Love the Avenger. By the author of “All for Gr...  A., A. A.   
3  Welsh Sketches, chiefly ecclesiastical, to the...  A., E. S.   
4  [The World in which I live, and my place in it...  A., E. S.   

                                   Contributors  Corporate Author  \
0                               FORBES, Walter.               NaN   
1  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
2  BLAZE DE BURY, Marie Pauline Rose - Baroness               NaN   
3                   Appleyard, Ernest Silvanus.               NaN   
4                           BROOME, John Henry.               NaN   

   Corporate Contributors Former owner  Engraver Issuance type  \
0                     NaN          NaN       NaN   monographic   
1                     NaN          NaN       NaN   monographic   
2                     NaN          NaN       NaN   monographic   
3                     NaN          NaN       NaN   monographic   
4                     NaN          NaN       NaN   monographic   

                                          Flickr URL  \
0  http://www.flickr.com/photos/britishlibrary/ta...   
1  http://www.flickr.com/photos/britishlibrary/ta...   
2  http://www.flickr.com/photos/britishlibrary/ta...   
3  http://www.flickr.com/photos/britishlibrary/ta...   
4  http://www.flickr.com/photos/britishlibrary/ta...   

                            Shelfmarks  
0    British Library HMNTS 12641.b.30.  
1    British Library HMNTS 12626.cc.2.  
2    British Library HMNTS 12625.dd.1.  
3  British Library HMNTS 10369.bbb.15.  
4     British Library HMNTS 9007.d.28.

In [212]:
fliker.shape #je regarde la taille du dataset

(8287, 15)

In [213]:
fliker.columns #je regarde les colonnes

Index(['Identifier', 'Edition Statement', 'Place of Publication',
       'Date of Publication', 'Publisher', 'Title', 'Author', 'Contributors',
       'Corporate Author', 'Corporate Contributors', 'Former owner',
       'Engraver', 'Issuance type', 'Flickr URL', 'Shelfmarks'],
      dtype='object')

In [214]:
fliker.dtypes #je regarde les types et si ils correspondent aux valeurs des colonnes

Identifier                  int64
Edition Statement          object
Place of Publication       object
Date of Publication        object
Publisher                  object
Title                      object
Author                     object
Contributors               object
Corporate Author          float64
Corporate Contributors    float64
Former owner               object
Engraver                  float64
Issuance type              object
Flickr URL                 object
Shelfmarks                 object
dtype: object

In [215]:
fliker.isnull().sum() #je regarde si il y a des valeurs manquantes

Identifier                   0
Edition Statement         7514
Place of Publication         0
Date of Publication        181
Publisher                 4195
Title                        0
Author                    1778
Contributors                 0
Corporate Author          8287
Corporate Contributors    8287
Former owner              8286
Engraver                  8287
Issuance type                0
Flickr URL                   0
Shelfmarks                   0
dtype: int64

In [216]:
#Je vois que pour les colonnes 'Corporate Author', 'Corporate Contributors', 'Former owner', 'Engraver','Edition Statement' PRESQUE toutes les colonnes sont vides donc je les supprime

fliker.drop(['Corporate Author', 'Corporate Contributors', 'Former owner', 'Engraver','Edition Statement'],  axis=1,inplace=True)
fliker

Identifier      Place of Publication Date of Publication  \
0            206                    London         1879 [1878]   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8282     4158088                    London                1838   
8283     4158128                     Derby            1831, 32   
8284     4159563                    London           [1806]-22   
8285     4159587       Newcastle upon Tyne                1834   
8286     4160339                    London             1834-43   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   
8286                      NaN   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   
8286  Collectanea Topographica et Genealogica. [Firs...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   
8286                                                NaN   

                                           Contributors Issuance type  \
0                                       FORBES, Walter.   monographic   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3                           Appleyard, Ernest Silvanus.   monographic   
4                                   BROOME, John Henry.   monographic   
...                                                 ...           ...   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...   monographic   
8283                                     NOBLE, Thomas.   monographic   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S   monographic   
8285                               ROSS, M. - of Durham   monographic   
8286            BANDINEL, Bulkeley.|Nichols, John Gough    continuing   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8282  http://www.flickr.com/photos/britishlibrary/ta...   
8283  http://www.flickr.com/photos/britishlibrary

In [217]:
fliker.isnull().sum()

Identifier                 0
Place of Publication       0
Date of Publication      181
Publisher               4195
Title                      0
Author                  1778
Contributors               0
Issuance type              0
Flickr URL                 0
Shelfmarks                 0
dtype: int64

Nous allons étudier les pourcentage de nan values dans des lignes

In [218]:

#renvoie les lignes ayant des valeurs manquantes
rows_with_missing_values = fliker[fliker.isnull().any(axis=1)]

#renvoie les pourcentages de valeurs manquantes par ligne
missing_percentage_per_row = fliker.isnull().sum(axis=1) / len(fliker.columns) * 100
missing_percentage_per_row





0        0.0
1        0.0
2        0.0
3        0.0
4        0.0
        ... 
8282    10.0
8283     0.0
8284     0.0
8285     0.0
8286    20.0
Length: 8287, dtype: float64

In [219]:
rows_with_missing_values.shape #je regarde la taille du dataset

(5225, 10)

In [220]:
missing_percentage_per_row_9percent = missing_percentage_per_row[missing_percentage_per_row > 9]
missing_percentage_per_row_9percent.shape

(5225,)

In [221]:
missing_percentage_per_row_10percent = missing_percentage_per_row[missing_percentage_per_row > 10]
missing_percentage_per_row_10percent.shape

(877,)

In [222]:
missing_percentage_per_row_10percent

7       20.0
13      20.0
35      20.0
39      20.0
61      20.0
        ... 
8267    20.0
8268    30.0
8270    20.0
8272    30.0
8286    20.0
Length: 877, dtype: float64

JE VOIS QUE TOUTES LES LIGNES AYANT DES NAN VALUES ONT UN POURCENTAGE SUPERIEUR STRICT A 9%

En revanche il n'y en a que 877 qui en ont avec un pourcentage supérieur strict à 10%

Nous allons drop ces lignes

In [223]:
fliker.drop(missing_percentage_per_row_10percent.index, inplace = True) #je drop les lignes qui ont plus de 10% de valeurs manquantes
fliker

Identifier      Place of Publication Date of Publication  \
0            206                    London         1879 [1878]   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8281     4157862       Newcastle-upon-Tyne                1867   
8282     4158088                    London                1838   
8283     4158128                     Derby            1831, 32   
8284     4159563                    London           [1806]-22   
8285     4159587       Newcastle upon Tyne                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8281               T. Fordyce   
8282                      NaN   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8281  Local Records; or, Historical Register of rema...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8281      FORDYCE, T. - Printer, of Newcastle-upon-Tyne   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   

                                           Contributors Issuance type  \
0                                       FORBES, Walter.   monographic   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3                           Appleyard, Ernest Silvanus.   monographic   
4                                   BROOME, John Henry.   monographic   
...                                                 ...           ...   
8281   SYKES, John - Bookseller, of Newcastle-upon-Tyne   monographic   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...   monographic   
8283                                     NOBLE, Thomas.   monographic   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S   monographic   
8285                               ROSS, M. - of Durham   monographic   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8281  http://www.flickr.com/photos/britishlibrary/ta...   
8282  http://www.flickr.com/photos/britishlibrary

In [224]:
fliker.isnull().sum()

Identifier                 0
Place of Publication       0
Date of Publication        2
Publisher               3318
Title                      0
Author                  1028
Contributors               0
Issuance type              0
Flickr URL                 0
Shelfmarks                 0
dtype: int64

Pour les nan values restante je vais les remplacer pour Unknown

In [225]:
fliker.fillna("Unknown", inplace=True)

In [226]:
fliker.isnull().sum()

Identifier              0
Place of Publication    0
Date of Publication     0
Publisher               0
Title                   0
Author                  0
Contributors            0
Issuance type           0
Flickr URL              0
Shelfmarks              0
dtype: int64

In [227]:
#On prends seulement les 4 premiers caractères de la colonne 'Date of Publication'
fliker.loc[fliker['Date of Publication'].str.len() > 4, 'Date of Publication'] = fliker['Date of Publication'].str[:4]
fliker

Identifier      Place of Publication Date of Publication  \
0            206                    London                1879   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8281     4157862       Newcastle-upon-Tyne                1867   
8282     4158088                    London                1838   
8283     4158128                     Derby                1831   
8284     4159563                    London                [180   
8285     4159587       Newcastle upon Tyne                1834   

                    Publisher  \
0            S. Tinsley & Co.   
1                Virtue & Co.   
2       Bradbury, Evans & Co.   
3               James Darling   
4        Wertheim & Macintosh   
...                       ...   
8281               T. Fordyce   
8282                  Unknown   
8283          M. Mozley & Son   
8284  T. Cadell and W. Davies   
8285         Mackenzie & Dent   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8281  Local Records; or, Historical Register of rema...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8284  Magna Britannia; being a concise topographical...   
8285  An historical, topographical and descriptive v...   

                                                 Author  \
0                                                 A. A.   
1                                             A., A. A.   
2                                             A., A. A.   
3                                             A., E. S.   
4                                             A., E. S.   
...                                                 ...   
8281      FORDYCE, T. - Printer, of Newcastle-upon-Tyne   
8282                 GIDDY, afterwards GILBERT, Davies.   
8283                         GLOVER, Stephen - of Derby   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   
8285                              Mackenzie, E. (Eneas)   

                                           Contributors Issuance type  \
0                                       FORBES, Walter.   monographic   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3                           Appleyard, Ernest Silvanus.   monographic   
4                                   BROOME, John Henry.   monographic   
...                                                 ...           ...   
8281   SYKES, John - Bookseller, of Newcastle-upon-Tyne   monographic   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...   monographic   
8283                                     NOBLE, Thomas.   monographic   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S   monographic   
8285                               ROSS, M. - of Durham   monographic   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8281  http://www.flickr.com/photos/britishlibrary/ta...   
8282  http://www.flickr.com/photos/britishlibrary

In [228]:
import re

# Filtrer les lignes où "Date of Publication" ne contient pas exactement 4 chiffres
lignes_date_publi_bizarre = fliker[~fliker["Date of Publication"].str.match(r'^\d{4}$', na=False)]
lignes_date_publi_bizarre

Identifier Place of Publication Date of Publication  \
24          5385               London                [189   
25          5389               London                [189   
34         11361              Leipzig                [189   
37         13364               London                [188   
38         14466               London                [186   
...          ...                  ...                 ...   
8223     4002305             New York                [187   
8225     4003256            Sheffield                [185   
8227     4006300             Augsburg                [186   
8265     4112839               London                [184   
8284     4159563               London                [180   

                    Publisher  \
24      Bradbury, Agnew & Co.   
25      Bradbury, Agnew & Co.   
34                    Unknown   
37     Griffith, Farran & Co.   
38               Adams & King   
...                       ...   
8223    American News Company   
8225          Ridge & Jackson   
8227                  Unknown   
8265         Willoughby & Co.   
8284  T. Cadell and W. Davies   

                                                  Title  \
24    [The comic history of England ... With twenty ...   
25    [The Comic History of Rome ... Illustrated by ...   
34    Tirol und Vorarlberg ... Mit ... einer Karte, ...   
37    A Hand-book and History of Sidmouth from the T...   
38    The Historie of Eald Street, now called Old St...   
...                                                 ...   
8223  Round the world with General Grant. A narrativ...   
8225  The Real Aspect of the Trevor Controversy. The...   
8227  Die Bayern im deutschen Kriege von 1866 ... Er...   
8265  The Vicar of Wakefield ... And a prefatory mem...   
8284  Magna Britannia; being a concise topographical...   

                                                 Author  \
24                           A'BECKETT, Gilbert Abbott.   
25                           A'BECKETT, Gilbert Abbott.   
34                    ACHLEITNER, Arthur - and UBL (E.)   
37                                              Unknown   
38                                              Unknown   
...                                                 ...   
8223                               YOUNG, John Russell.   
8225                                    YOUNGE, Robert.   
8227                                          Z., A. v.   
8265                                  Goldsmith, Oliver   
8284  LYSONS, Daniel - M.A., F.R.S., and LYSONS (Sam...   

                                           Contributors Issuance type  \
24                                 LEECH, John - Artist   monographic   
25                                 LEECH, John - Artist   monographic   
34                                              UBL, E.   monographic   
37                           HUTCHINSON, Peter Orlando.   monographic   
38                    KING - Printer, of Goswell Street   monographic   
...                                                 ...           ...   
8223  GRANT, Ulysses Simpson - President of the Unit...   monographic   
8225                                    TREVOR, George.   monographic   
8227                                    ZERZOG, A. von.   monographic   
8265                               BUSSEY, George Moir.   monographic   
8284           GREGSON, Matthew.|LYSONS, Samuel - F.R.S   monographic   

                                             Flickr URL  \
24    http://www.flickr.com/photos/britishlibrary/ta...   
25    http://www.flickr.com/photos/britishlibrary/ta...   
34    http://www.flickr.com/photos/britishlibrary/ta...   
37    http://www.flickr.com/photos/britishlibrary/ta...   
38    http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8223  http://www.flickr.com/photos/britishlibrary/ta...   
8225  http://www.flickr.com/photos/britishlibrary/ta...   
8227  http://www.flickr.com/photos/britishlibrary/

In [229]:
#je drop de fliker toutes les dates bizarres (ne contenant pas 4 chiffres)
fliker.drop(lignes_date_publi_bizarre.index, inplace=True)
fliker

Identifier      Place of Publication Date of Publication  \
0            206                    London                1879   
1            216  London; Virtue & Yorston                1868   
2            218                    London                1869   
3            472                    London                1851   
4            480                    London                1857   
...          ...                       ...                 ...   
8280     4157746                    London                1811   
8281     4157862       Newcastle-upon-Tyne                1867   
8282     4158088                    London                1838   
8283     4158128                     Derby                1831   
8285     4159587       Newcastle upon Tyne                1834   

                   Publisher  \
0           S. Tinsley & Co.   
1               Virtue & Co.   
2      Bradbury, Evans & Co.   
3              James Darling   
4       Wertheim & Macintosh   
...                      ...   
8280  F. C. and J. Rivington   
8281              T. Fordyce   
8282                 Unknown   
8283         M. Mozley & Son   
8285        Mackenzie & Dent   

                                                  Title  \
0                     Walter Forbes. [A novel.] By A. A   
1     All for Greed. [A novel. The dedication signed...   
2     Love the Avenger. By the author of “All for Gr...   
3     Welsh Sketches, chiefly ecclesiastical, to the...   
4     [The World in which I live, and my place in it...   
...                                                 ...   
8280  The New Chronicles of England and France, in t...   
8281  Local Records; or, Historical Register of rema...   
8282  The Parochial History of Cornwall, founded on,...   
8283  The History and Gazetteer of the County of Der...   
8285  An historical, topographical and descriptive v...   

                                             Author  \
0                                             A. A.   
1                                         A., A. A.   
2                                         A., A. A.   
3                                         A., E. S.   
4                                         A., E. S.   
...                                             ...   
8280                                FABYAN, Robert.   
8281  FORDYCE, T. - Printer, of Newcastle-upon-Tyne   
8282             GIDDY, afterwards GILBERT, Davies.   
8283                     GLOVER, Stephen - of Derby   
8285                          Mackenzie, E. (Eneas)   

                                           Contributors Issuance type  \
0                                       FORBES, Walter.   monographic   
1          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
2          BLAZE DE BURY, Marie Pauline Rose - Baroness   monographic   
3                           Appleyard, Ernest Silvanus.   monographic   
4                                   BROOME, John Henry.   monographic   
...                                                 ...           ...   
8280  ELLIS, Henry - Sir, Principal Librarian of the...   monographic   
8281   SYKES, John - Bookseller, of Newcastle-upon-Tyne   monographic   
8282  BOASE, Henry Samuel.|HALS, William.|LYSONS, Da...   monographic   
8283                                     NOBLE, Thomas.   monographic   
8285                               ROSS, M. - of Durham   monographic   

                                             Flickr URL  \
0     http://www.flickr.com/photos/britishlibrary/ta...   
1     http://www.flickr.com/photos/britishlibrary/ta...   
2     http://www.flickr.com/photos/britishlibrary/ta...   
3     http://www.flickr.com/photos/britishlibrary/ta...   
4     http://www.flickr.com/photos/britishlibrary/ta...   
...                                                 ...   
8280  http://www.flickr.com/photos/britishlibrary/ta...   
8281  http://www.flickr.com/photos/britishlibrary/ta...   
8282  http://www.flickr.com/photos/britishlibrary/

# Exercice 7 :

In [230]:
initialData = pd.read_csv('melb_data.csv', delimiter= ',')
initialData.head()

Unnamed: 0      Suburb           Address  Rooms Type      Price Method  \
0           1  Abbotsford      85 Turner St      2    h  1480000.0      S   
1           2  Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2           4  Abbotsford      5 Charles St      3    h  1465000.0     SP   
3           5  Abbotsford  40 Federation La      3    h   850000.0     PI   
4           6  Abbotsford       55a Park St      4    h  1600000.0     VB   

  SellerG       Date  Distance  ...  Bathroom  Car  Landsize  BuildingArea  \
0  Biggin  3/12/2016       2.5  ...       1.0  1.0     202.0           NaN   
1  Biggin  4/02/2016       2.5  ...       1.0  0.0     156.0          79.0   
2  Biggin  4/03/2017       2.5  ...       2.0  0.0     134.0         150.0   
3  Biggin  4/03/2017       2.5  ...       2.0  1.0      94.0           NaN   
4  Nelson  4/06/2016       2.5  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea  Lattitude Longtitude             Regionname  \
0        NaN        Yarra   -37.7996   144.9984  Northern Metropolitan   
1     1900.0        Yarra   -37.8079   144.9934  Northern Metropolitan   
2     1900.0        Yarra   -37.8093   144.9944  Northern Metropolitan   
3        NaN        Yarra   -37.7969   144.9969  Northern Metropolitan   
4     2014.0        Yarra   -37.8072   144.9941  Northern Metropolitan   

   Propertycount  
0         4019.0  
1         4019.0  
2         4019.0  
3         4019.0  
4         4019.0  

[5 rows x 22 columns]

In [231]:
initialData.columns

Index(['Unnamed: 0', 'Suburb', 'Address', 'Rooms', 'Type', 'Price', 'Method',
       'SellerG', 'Date', 'Distance', 'Postcode', 'Bedroom2', 'Bathroom',
       'Car', 'Landsize', 'BuildingArea', 'YearBuilt', 'CouncilArea',
       'Lattitude', 'Longtitude', 'Regionname', 'Propertycount'],
      dtype='object')

In [232]:
initialData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18396 entries, 0 to 18395
Data columns (total 22 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Unnamed: 0     18396 non-null  int64  
 1   Suburb         18396 non-null  object 
 2   Address        18396 non-null  object 
 3   Rooms          18396 non-null  int64  
 4   Type           18396 non-null  object 
 5   Price          18396 non-null  float64
 6   Method         18396 non-null  object 
 7   SellerG        18396 non-null  object 
 8   Date           18396 non-null  object 
 9   Distance       18395 non-null  float64
 10  Postcode       18395 non-null  float64
 11  Bedroom2       14927 non-null  float64
 12  Bathroom       14925 non-null  float64
 13  Car            14820 non-null  float64
 14  Landsize       13603 non-null  float64
 15  BuildingArea   7762 non-null   float64
 16  YearBuilt      8958 non-null   float64
 17  CouncilArea    12233 non-null  object 
 18  Lattit

In [233]:
initialData.shape

(18396, 22)

In [234]:
initialData.isnull().sum() #Total missing values par colonnes

Unnamed: 0           0
Suburb               0
Address              0
Rooms                0
Type                 0
Price                0
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          3469
Bathroom          3471
Car               3576
Landsize          4793
BuildingArea     10634
YearBuilt         9438
CouncilArea       6163
Lattitude         3332
Longtitude        3332
Regionname           1
Propertycount        1
dtype: int64

In [235]:
initialData.isnull().sum(axis=1) #Total missing values par lignes

0        2
1        0
2        0
3        2
4        0
        ..
18391    2
18392    1
18393    4
18394    1
18395    2
Length: 18396, dtype: int64

### Gestion valeurs manquantes par colonnes 

In [236]:
def missing_colums(originDB):
  colsWithMissing = []

  for col in originDB.columns:
    if originDB[col].isnull().any():
      colsWithMissing.append(col)

  reduced_original_data = originDB.drop(colsWithMissing, axis=1)

  return colsWithMissing, reduced_original_data


In [237]:
colsWithMissing, reduced_original_data = missing_colums(initialData)

In [238]:
colsWithMissing #je regarde les colonnes qui ont des valeurs manquantes

['Distance',
 'Postcode',
 'Bedroom2',
 'Bathroom',
 'Car',
 'Landsize',
 'BuildingArea',
 'YearBuilt',
 'CouncilArea',
 'Lattitude',
 'Longtitude',
 'Regionname',
 'Propertycount']

In [239]:
reduced_original_data #je regarde le dataset sans les colonnes qui ont des valeurs manquantes

Unnamed: 0        Suburb           Address  Rooms Type      Price  \
0               1    Abbotsford      85 Turner St      2    h  1480000.0   
1               2    Abbotsford   25 Bloomburg St      2    h  1035000.0   
2               4    Abbotsford      5 Charles St      3    h  1465000.0   
3               5    Abbotsford  40 Federation La      3    h   850000.0   
4               6    Abbotsford       55a Park St      4    h  1600000.0   
...           ...           ...               ...    ...  ...        ...   
18391       23540  Williamstown   8/2 Thompson St      2    t   622500.0   
18392       23541  Williamstown      96 Verdon St      4    h  2500000.0   
18393       23544     Yallambie      17 Amaroo Wy      4    h  1100000.0   
18394       23545    Yarraville        6 Agnes St      4    h  1285000.0   
18395       23546    Yarraville     33 Freeman St      4    h  1050000.0   

      Method     SellerG        Date  
0          S      Biggin   3/12/2016  
1          S      Biggin   4/02/2016  
2         SP      Biggin   4/03/2017  
3         PI      Biggin   4/03/2017  
4         VB      Nelson   4/06/2016  
...      ...         ...         ...  
18391     SP        Greg  26/08/2017  
18392     PI     Sweeney  26/08/2017  
18393      S  Buckingham  26/08/2017  
18394     SP     Village  26/08/2017  
18395     VB     Village  26/08/2017  

[18396 rows x 9 columns]

In [240]:
# Calculer le pourcentage de valeurs manquantes par colonne
missing_percentage_per_column = initialData.isnull().sum() / len(initialData) * 100

# Afficher les pourcentages de valeurs manquantes pour chaque colonne
missing_percentage_per_column


Unnamed: 0        0.000000
Suburb            0.000000
Address           0.000000
Rooms             0.000000
Type              0.000000
Price             0.000000
Method            0.000000
SellerG           0.000000
Date              0.000000
Distance          0.005436
Postcode          0.005436
Bedroom2         18.857360
Bathroom         18.868232
Car              19.439008
Landsize         26.054577
BuildingArea     57.806045
YearBuilt        51.304631
CouncilArea      33.501848
Lattitude        18.112633
Longtitude       18.112633
Regionname        0.005436
Propertycount     0.005436
dtype: float64

### Gestion valeurs manquantes par lignes

In [241]:
def missing_rows(originDB):
  rows_with_missing_values = originDB[originDB.isnull().any(axis=1)] #.any(axis=1)  pour vérifier si au moins une valeur manquante est présente dans chaque ligne

  reduced_original_data = originDB.drop(rows_with_missing_values.index, axis=0)

  return rows_with_missing_values, reduced_original_data



In [242]:
rows_with_missing_values, reduced_original_data_row = missing_rows(initialData)

In [243]:
missing_percentage_per_row = initialData.isnull().sum(axis=1) / len(initialData.columns) * 100

# Afficher les pourcentages de valeurs manquantes pour chaque ligne
missing_percentage_per_row

#on met axis=1 pour avoir la somme pour chaque element de la ligne pour CHAQUE COLONNE


0         9.090909
1         0.000000
2         0.000000
3         9.090909
4         0.000000
           ...    
18391     9.090909
18392     4.545455
18393    18.181818
18394     4.545455
18395     9.090909
Length: 18396, dtype: float64

In [244]:
missing_percentage_per_row[missing_percentage_per_row > 5] #je garde les lignes qui ont plus de 5% de valeurs manquantes

0         9.090909
3         9.090909
5         9.090909
8        40.909091
9        40.909091
           ...    
18388     9.090909
18390     9.090909
18391     9.090909
18393    18.181818
18395     9.090909
Length: 10064, dtype: float64

In [245]:
missing_percentage_per_row_5 = missing_percentage_per_row[missing_percentage_per_row > 5] 
new_Data = initialData.drop(missing_percentage_per_row_5.index) #je drop les lignes qui ont plus de 5% de valeurs manquantes
new_Data

Unnamed: 0        Suburb          Address  Rooms Type      Price  \
1               2    Abbotsford  25 Bloomburg St      2    h  1035000.0   
2               4    Abbotsford     5 Charles St      3    h  1465000.0   
4               6    Abbotsford      55a Park St      4    h  1600000.0   
6              11    Abbotsford     124 Yarra St      3    h  1876000.0   
7              14    Abbotsford    98 Charles St      2    h  1636000.0   
...           ...           ...              ...    ...  ...        ...   
18382       23527      Watsonia    76 Kenmare St      2    h   650000.0   
18384       23532      Werribee      5 Nuragi Ct      4    h   635000.0   
18389       23538  Williamstown    77 Merrett Dr      3    h  1031000.0   
18392       23541  Williamstown     96 Verdon St      4    h  2500000.0   
18394       23545    Yarraville       6 Agnes St      4    h  1285000.0   

      Method        SellerG        Date  Distance  ...  Bathroom  Car  \
1          S         Biggin   4/02/2016       2.5  ...       1.0  0.0   
2         SP         Biggin   4/03/2017       2.5  ...       2.0  0.0   
4         VB         Nelson   4/06/2016       2.5  ...       1.0  2.0   
6          S         Nelson   7/05/2016       2.5  ...       2.0  0.0   
7          S         Nelson   8/10/2016       2.5  ...       1.0  2.0   
...      ...            ...         ...       ...  ...       ...  ...   
18382     PI       Morrison  26/08/2017      14.5  ...       1.0  1.0   
18384      S  hockingstuart  26/08/2017      14.7  ...       2.0  1.0   
18389     SP       Williams  26/08/2017       6.8  ...       2.0  2.0   
18392     PI        Sweeney  26/08/2017       6.8  ...       1.0  5.0   
18394     SP        Village  26/08/2017       6.3  ...       1.0  1.0   

       Landsize  BuildingArea  YearBuilt  CouncilArea  Lattitude Longtitude  \
1         156.0          79.0     1900.0        Yarra  -37.80790  144.99340   
2         134.0         150.0     1900.0        Yarra  -37.80930  144.99440   
4         120.0         142.0     2014.0        Yarra  -37.80720  144.99410   
6         245.0         210.0     1910.0        Yarra  -37.80240  144.99930   
7         256.0         107.0     1890.0        Yarra  -37.80600  144.99540   
...         ...           ...        ...          ...        ...        ...   
18382     210.0          79.0     2006.0          NaN  -37.70657  145.07878   
18384     662.0         172.0     1980.0          NaN  -37.89327  144.64789   
18389     333.0         133.0     1995.0          NaN  -37.85927  144.87904   
18392     866.0         157.0     1920.0          NaN  -37.85908  144.89299   
18394     362.0         112.0     1920.0          NaN  -37.81188  144.88449   

                  Regionname  Propertycount  
1      Northern Metropolitan         4019.0  
2      Northern Metropolitan         4019.0  
4      Northern Metropolitan         4019.0  
6      Northern Metropolitan         4019.0  
7      Northern Metropolitan         4019.0  
...                      ...            ...  
18382  Northern Metropolitan         2329.0  
18384   Western Metropolitan        16166.0  
18389   Western Metropolitan         6380.0  
18392   Western Metropolitan         6380.0  
18394   Western Metropolitan         6543.0  

[8332 rows x 22 columns]

In [246]:
colsWithMissing_new_data, reduced_new_data = missing_colums(new_Data) #je refais la fonction pour voir les colonnes qui ont des valeurs manquantes

In [247]:
colsWithMissing_new_data #je regarde les colonnes qui ont des valeurs manquantes POUR LE NOUVEAU DATASET

['BuildingArea', 'YearBuilt', 'CouncilArea']

In [248]:
colsWithMissing #je regarde les colonnes qui ont des valeurs manquantes POUR LE DATASET INITIAL

['Distance',
 'Postcode',
 'Bedroom2',
 'Bathroom',
 'Car',
 'Landsize',
 'BuildingArea',
 'YearBuilt',
 'CouncilArea',
 'Lattitude',
 'Longtitude',
 'Regionname',
 'Propertycount']

En ayant supprimer d'abord les lignes avec les missing values plutot que les colonnes cela m'a permis de garder des colonnes car celle ci contenant les lignes qu'on a supprimé

In [249]:
new_Data.isnull().sum()

Unnamed: 0          0
Suburb              0
Address             0
Rooms               0
Type                0
Price               0
Method              0
SellerG             0
Date                0
Distance            0
Postcode            0
Bedroom2            0
Bathroom            0
Car                 0
Landsize            0
BuildingArea     1252
YearBuilt         250
CouncilArea       634
Lattitude           0
Longtitude          0
Regionname          0
Propertycount       0
dtype: int64

In [250]:
new_Data.loc[:,'Price'].describe()
new_Data.loc[:,'Price'].fillna(new_Data['Price'].mean(),inplace=True)

# Exercice 8 :

In [251]:
from sklearn.impute import SimpleImputer
my_imputer = SimpleImputer(strategy='mean') #Tu remplaces les Nan par ce que tu mets dans la strategy (ici c'est la moyenne)


In [252]:
#je fais une copie du dataset avec les valeurs numériques

origin_Data_Num = initialData[["Rooms", "Price", "Distance", "Postcode", "Bedroom2", "Bathroom", "Car", "Landsize", "BuildingArea", "YearBuilt", "Lattitude", "Longtitude", "Propertycount"]]

origin_Data_Num


Rooms      Price  Distance  Postcode  Bedroom2  Bathroom  Car  \
0          2  1480000.0       2.5    3067.0       2.0       1.0  1.0   
1          2  1035000.0       2.5    3067.0       2.0       1.0  0.0   
2          3  1465000.0       2.5    3067.0       3.0       2.0  0.0   
3          3   850000.0       2.5    3067.0       3.0       2.0  1.0   
4          4  1600000.0       2.5    3067.0       3.0       1.0  2.0   
...      ...        ...       ...       ...       ...       ...  ...   
18391      2   622500.0       6.8    3016.0       2.0       2.0  1.0   
18392      4  2500000.0       6.8    3016.0       4.0       1.0  5.0   
18393      4  1100000.0      12.7    3085.0       4.0       3.0  2.0   
18394      4  1285000.0       6.3    3013.0       4.0       1.0  1.0   
18395      4  1050000.0       6.3    3013.0       4.0       2.0  2.0   

       Landsize  BuildingArea  YearBuilt  Lattitude  Longtitude  Propertycount  
0         202.0           NaN        NaN  -37.79960   144.99840         4019.0  
1         156.0          79.0     1900.0  -37.80790   144.99340         4019.0  
2         134.0         150.0     1900.0  -37.80930   144.99440         4019.0  
3          94.0           NaN        NaN  -37.79690   144.99690         4019.0  
4         120.0         142.0     2014.0  -37.80720   144.99410         4019.0  
...         ...           ...        ...        ...         ...            ...  
18391       NaN          89.0     2010.0  -37.86393   144.90484         6380.0  
18392     866.0         157.0     1920.0  -37.85908   144.89299         6380.0  
18393       NaN           NaN        NaN  -37.72006   145.10547         1369.0  
18394     362.0         112.0     1920.0  -37.81188   144.88449         6543.0  
18395       NaN         139.0     1950.0  -37.81829   144.87404         6543.0  

[18396 rows x 13 columns]

In [253]:

my_imputer.fit(origin_Data_Num)

SimpleImputer()

In [254]:
origin_Data_Num.isnull().sum()

Rooms                0
Price                0
Distance             1
Postcode             1
Bedroom2          3469
Bathroom          3471
Car               3576
Landsize          4793
BuildingArea     10634
YearBuilt         9438
Lattitude         3332
Longtitude        3332
Propertycount        1
dtype: int64

In [255]:
my_imputer.statistics_
Transform_Data_num = my_imputer.transform(origin_Data_Num)
df_Transform_Data_num = pd.DataFrame(Transform_Data_num, columns=origin_Data_Num.columns, index=origin_Data_Num.index)




In [256]:
df_Transform_Data_num

Rooms      Price  Distance  Postcode  Bedroom2  Bathroom  Car  \
0        2.0  1480000.0       2.5    3067.0       2.0       1.0  1.0   
1        2.0  1035000.0       2.5    3067.0       2.0       1.0  0.0   
2        3.0  1465000.0       2.5    3067.0       3.0       2.0  0.0   
3        3.0   850000.0       2.5    3067.0       3.0       2.0  1.0   
4        4.0  1600000.0       2.5    3067.0       3.0       1.0  2.0   
...      ...        ...       ...       ...       ...       ...  ...   
18391    2.0   622500.0       6.8    3016.0       2.0       2.0  1.0   
18392    4.0  2500000.0       6.8    3016.0       4.0       1.0  5.0   
18393    4.0  1100000.0      12.7    3085.0       4.0       3.0  2.0   
18394    4.0  1285000.0       6.3    3013.0       4.0       1.0  1.0   
18395    4.0  1050000.0       6.3    3013.0       4.0       2.0  2.0   

         Landsize  BuildingArea    YearBuilt  Lattitude  Longtitude  \
0      202.000000    151.220219  1965.879996  -37.79960   144.99840   
1      156.000000     79.000000  1900.000000  -37.80790   144.99340   
2      134.000000    150.000000  1900.000000  -37.80930   144.99440   
3       94.000000    151.220219  1965.879996  -37.79690   144.99690   
4      120.000000    142.000000  2014.000000  -37.80720   144.99410   
...           ...           ...          ...        ...         ...   
18391  558.116371     89.000000  2010.000000  -37.86393   144.90484   
18392  866.000000    157.000000  1920.000000  -37.85908   144.89299   
18393  558.116371    151.220219  1965.879996  -37.72006   145.10547   
18394  362.000000    112.000000  1920.000000  -37.81188   144.88449   
18395  558.116371    139.000000  1950.000000  -37.81829   144.87404   

       Propertycount  
0             4019.0  
1             4019.0  
2             4019.0  
3             4019.0  
4             4019.0  
...              ...  
18391         6380.0  
18392         6380.0  
18393         1369.0  
18394         6543.0  
18395         6543.0  

[18396 rows x 13 columns]

In [257]:
df_Transform_Data_num.isnull().sum()

Rooms            0
Price            0
Distance         0
Postcode         0
Bedroom2         0
Bathroom         0
Car              0
Landsize         0
BuildingArea     0
YearBuilt        0
Lattitude        0
Longtitude       0
Propertycount    0
dtype: int64

# Exercice 9 :

Je vais me baser sur le dataset 'melb_data.csv'

In [258]:
melb_data = pd.read_csv('melb_data.csv', delimiter= ',')
melb_data.head()

Unnamed: 0      Suburb           Address  Rooms Type      Price Method  \
0           1  Abbotsford      85 Turner St      2    h  1480000.0      S   
1           2  Abbotsford   25 Bloomburg St      2    h  1035000.0      S   
2           4  Abbotsford      5 Charles St      3    h  1465000.0     SP   
3           5  Abbotsford  40 Federation La      3    h   850000.0     PI   
4           6  Abbotsford       55a Park St      4    h  1600000.0     VB   

  SellerG       Date  Distance  ...  Bathroom  Car  Landsize  BuildingArea  \
0  Biggin  3/12/2016       2.5  ...       1.0  1.0     202.0           NaN   
1  Biggin  4/02/2016       2.5  ...       1.0  0.0     156.0          79.0   
2  Biggin  4/03/2017       2.5  ...       2.0  0.0     134.0         150.0   
3  Biggin  4/03/2017       2.5  ...       2.0  1.0      94.0           NaN   
4  Nelson  4/06/2016       2.5  ...       1.0  2.0     120.0         142.0   

   YearBuilt  CouncilArea  Lattitude Longtitude             Regionname  \
0        NaN        Yarra   -37.7996   144.9984  Northern Metropolitan   
1     1900.0        Yarra   -37.8079   144.9934  Northern Metropolitan   
2     1900.0        Yarra   -37.8093   144.9944  Northern Metropolitan   
3        NaN        Yarra   -37.7969   144.9969  Northern Metropolitan   
4     2014.0        Yarra   -37.8072   144.9941  Northern Metropolitan   

   Propertycount  
0         4019.0  
1         4019.0  
2         4019.0  
3         4019.0  
4         4019.0  

[5 rows x 22 columns]

In [259]:
melb_data.isnull().sum()

Unnamed: 0           0
Suburb               0
Address              0
Rooms                0
Type                 0
Price                0
Method               0
SellerG              0
Date                 0
Distance             1
Postcode             1
Bedroom2          3469
Bathroom          3471
Car               3576
Landsize          4793
BuildingArea     10634
YearBuilt         9438
CouncilArea       6163
Lattitude         3332
Longtitude        3332
Regionname           1
Propertycount        1
dtype: int64

In [260]:
melb_data.dtypes

Unnamed: 0         int64
Suburb            object
Address           object
Rooms              int64
Type              object
Price            float64
Method            object
SellerG           object
Date              object
Distance         float64
Postcode         float64
Bedroom2         float64
Bathroom         float64
Car              float64
Landsize         float64
BuildingArea     float64
YearBuilt        float64
CouncilArea       object
Lattitude        float64
Longtitude       float64
Regionname        object
Propertycount    float64
dtype: object

In [261]:
numerical_cols = melb_data.select_dtypes(exclude=['object']).columns
categorical_cols = melb_data.select_dtypes(exclude=['number']).columns

print("Numerical columns:", numerical_cols)
print("Categorical columns:", categorical_cols)

Numerical columns: Index(['Unnamed: 0', 'Rooms', 'Price', 'Distance', 'Postcode', 'Bedroom2',
       'Bathroom', 'Car', 'Landsize', 'BuildingArea', 'YearBuilt', 'Lattitude',
       'Longtitude', 'Propertycount'],
      dtype='object')
Categorical columns: Index(['Suburb', 'Address', 'Type', 'Method', 'SellerG', 'Date', 'CouncilArea',
       'Regionname'],
      dtype='object')


In [262]:
for col in categorical_cols:
    print(col + ":", melb_data[col].nunique())


Suburb: 330
Address: 18134
Type: 3
Method: 5
SellerG: 305
Date: 58
CouncilArea: 33
Regionname: 8


Etant donner que je vais faire du one-hot encoding dans la pipeline je vais devoir garder seulement les colonnes avec un nunique < 10 car sinon cela ne sera plus pertinent

In [263]:
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.decomposition import PCA
from sklearn.ensemble import RandomForestClassifier


# Trouver les colonnes catégorielles avec moins de 10 catégories pour le OneHotEncoder (car sinon au dessus ce n'est pas optimal)
onehot_columns = [col for col in categorical_cols if melb_data[col].nunique() < 10]

# Trouver les autres colonnes catégorielles pour OrdinalEncoder
ordinal_columns = [col for col in categorical_cols if col not in onehot_columns]

# Pipeline pour les colonnes numériques (ex: imputation)
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')), #remplace les valeurs manquantes par la median (souvent meilleur methode)
    ('scaler', StandardScaler()) #centre et reduit les données
])

# Pipeline pour les colonnes OneHot
onehot_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),  #remplace les valeurs manquantes par la valeur la plus fréquente
    ('onehot', OneHotEncoder(handle_unknown='ignore')) #crée des colonnes pour chaque catégorie et met 1 si la catégorie est présente et 0 sinon
])

# Pipeline pour les colonnes Ordinal
ordinal_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')), #remplace les valeurs manquantes par la valeur la plus fréquente
    ('ordinal', OrdinalEncoder()) #transforme les valeurs catégorielles en valeurs numériques
])

# Combiner le tout dans un ColumnTransformer
preprocessor = ColumnTransformer(
    transformers=[
        ('num', numeric_transformer, numerical_cols),
        ('onehot', onehot_transformer, onehot_columns),
        ('ordinal', ordinal_transformer, ordinal_columns)
    ]
)

# Pipeline final avec modèle de classification (exemple RandomForest)
model_pipeline = Pipeline(steps=[
    ('preprocessor', preprocessor),
    ('pca', PCA(n_components=4)), #réduit la dimensionnalité du dataset
    ('classifier', RandomForestClassifier())
])


"""

Ce n'est pas demandé mais par la suite il faudrait faire le train test split et fit le modèle

"""

"\n\nCe n'est pas demandé mais par la suite il faudrait faire le train test split et fit le modèle\n\n"